Yatima/ForLurkRepo/Eval.lean

```diff
@@ -0,0 +1,138 @@
+  | .prod => match v₁, v₂ with
+    | .lit (.num x), .lit (.num y) => return .lit $ .num (x * y)
+    | _, _ => throw "error: not a number"
+  | .quot => match v₁, v₂ with
+    | .lit (.num x), .lit (.num y) => return .lit $ .num (x * y⁻¹)
+    | _, _ => throw "error: not a number"
+  | .eq => match v₁, v₂ with
+    | .lit (.num x), .lit (.num y) =>
+      return if x == y then .lit .t else .lit .nil
+    | _, _ => throw "error: not a number"
+  | .nEq => return if v₁ == v₂ then .lit .t else .lit .nil
+
+def bind (body : Expr) (ns : List Name) (as : List Expr) :
+    EvalM (Expr × List Name) := do
+  let rec aux (acc : List (Name × Expr)) :
+      List Name → List Expr → EvalM ((List (Name × Expr)) × List Name)
+    | n::ns, a::as => aux ((n, a) :: acc) ns as
+    | [], _::_ => throw "too many arguments"
+    | ns, [] => return (acc, ns)
+  let (binds, ns') ← aux [] ns as
+  return (.letE binds body, ns')
+
+def isAtom (e : Expr) : Bool := match e with 
+  | .cons .. => false 
+  | _ => true
+
+partial def evalM (env : Env) : Expr → EvalM Value
+  | .lit lit => return .lit lit
+  | .sym n => match env.find? n with
+    | some v => return v
+    | none => throw s!"{n} not found"
+  | .ifE tst con alt => do
+    match ← evalM env tst with
+    | .lit .t => evalM env con
+    | .lit .nil => evalM env alt
+    | _ => throw "not a boolean"
+  | .lam formals body => return .lam formals body
+  | .letE bindings body => do
+    let env' ← bindings.foldlM (init := env)
+      fun acc (n, e) => return acc.insert n (← evalM acc e)
+    evalM env' body
+  | .letRecE bindings body => default
+  | .app fn args => do
+    match ← evalM env fn with
+    | .lam ns body =>
+      let (body', ns') ← bind body ns args
+      if ns'.isEmpty then evalM env body' else return .lam ns body'
+    | _ => throw "app function is not a lambda"
+  | .quote _ => unreachable! -- not used for debugging/testing
+  | .binaryOp op e₁ e₂ => do evalBinaryOp op (← evalM env e₁) (← evalM env e₂)
+  | .atom e => if isAtom e then return .lit .t else return .lit .nil
+  | .cons e₁ e₂ => return .cons (← evalM env e₁) (← evalM env e₂)
+  | .strcons e₁ e₂ => do match (← evalM env e₁), (← evalM env e₂) with 
+    | .lit (.char c), .lit (.str s) => return .lit (.str ⟨c :: s.data⟩)
+    | .lit (.char _), x => throw s!"expected string value, got\n {x.pprint}" 
+    | x, _ => throw s!"expected char value, got\n {x.pprint}" 
+  -- TODO: add String support; `car "abc" ==> 'a'` 
+  | .car e => do match (← evalM env e) with 
+    | .cons e₁ _ => return e₁ 
+    | _ => throw "not a cons"
+  -- TODO: add String support; `cdr "abc" ==> "bc"`
+  | .cdr e => do match (← evalM env e) with 
+    | .cons e₁ _ => return e₁ 
+    | _ => throw "not a cons"
+  | .emit e => do
+    let v ← evalM env e
+    IO.println v.pprint
+    pure v
+  | .begin es => evalM env $ es.reverse.headD $ .lit .nil
+  | .currEnv => return .env env.toList
+
+def ppEval (e : Expr) (env : Env := default) : IO Format :=
```

Comment:
I changed this back 😅 hopefully that's ok. In my head, "pretty print" is a description so this is "pretty printing eval".